contrib/pod-master/podmaster.go

```diff
@@ -0,0 +1,174 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// podmaster is a simple utility, it attempts to acquire and maintain a lease-lock from etcd using compare-and-swap.
+// if it is the master, it copies a source file into a destination file.  If it is not the master, it makes sure it is removed.
+//
+// typical usage is to copy a Pod manifest from a staging directory into the kubelet's directory, for example:
+//   podmaster --etcd-servers=http://127.0.0.1:4001 --key=scheduler --source-file=/kubernetes/kube-scheduler.manifest --dest-file=/manifests/kube-scheduler.manifest
+
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools"
+
+	"github.com/coreos/go-etcd/etcd"
+	"github.com/golang/glog"
+	"github.com/spf13/pflag"
+)
+
+type Config struct {
+	etcdServers string
+	key         string
+	whoami      string
+	ttl         uint64
+	src         string
+	dest        string
+	sleep       time.Duration
+}
+
+// runs the election loop. never returns.
+func leaseAndUpdateLoop(c *Config, etcdClient *etcd.Client) {
+	for {
+		master, err := acquireOrRenewLease(c, etcdClient)
+		if err != nil {
+			glog.Errorf("Error in master election: %v", err)
+			continue
+		}
+		if err := update(c, master); err != nil {
+			glog.Errorf("Error updating files: %v", err)
+		}
+		time.Sleep(c.sleep)
+	}
+}
+
+// acquireOrRenewLease either races to acquire a new master lease, or update the existing master's lease
+// returns true if we have the lease, and an error if one occurs.
+// TODO: use the master election utility once it is merged in.
+func acquireOrRenewLease(c *Config, etcdClient *etcd.Client) (bool, error) {
+	result, err := etcdClient.Get(c.key, false, false)
+	if tools.IsEtcdNotFound(err) {
+		// there is no current master, try to become master, create will fail if the key already exists
+		_, err := etcdClient.Create(c.key, c.whoami, c.ttl)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	} else if result.Node.Value == c.whoami {
+		glog.Infof("key already exists, we are the master (%s)", result.Node.Value)
+		// we extend our lease @ 1/2 of the existing TTL, this ensures the master doesn't flap around
+		if result.Node.Expiration.Sub(time.Now()) < time.Duration(c.ttl/2)*time.Second {
+			_, err := etcdClient.CompareAndSwap(c.key, c.whoami, c.ttl, c.whoami, result.Node.ModifiedIndex)
+			if err != nil {
+				return false, err
+			}
+		}
+		return true, nil
+	} else {
+		glog.Infof("key already exists, the master is %s, sleeping.", result.Node.Value)
+		return false, nil
+	}
+}
+
+// update enacts the policy, copying a file if we are the master, and it doesn't exist.
+// deleting a file if we aren't the master and it does.
+func update(c *Config, master bool) error {
+	exists, err := exists(c.dest)
+	if err != nil {
+		return err
+	}
+	switch {
+	case master && !exists:
+		return copyFile(c.src, c.dest)
```

Comment:
I was thinking the latter-- that way if you do want to update, you can (safely) update the source file. Otherwise to really update you have to update both, which is kinda weird. But a comment/TODO is enough to make me happy.